### PR TITLE
[Gradient checkpointing] Restore backwards compatibility until v5

### LINKED ISSUE #14388: Wav2Vec2 CUDA memory usage doubled in v4.11.3 compared to v4.10.3 with the same batch size
## Environment info
- `transformers` version: 4.11.3
- Platform: Linux-5.11.0-40-generic-x86_64-with-glibc2.29
- Python version: 3.8.10
- PyTorch version (GPU?): 1.8.1+cu111 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes, 3090
- Using distributed or parallel set-up in script?: No

### Who can help
@patrickvonplaten, @anton-l

## Information

When using Wav2vec2 the memory usage roughly doubles when going from Huggingface v4.10.3 to v4.11.3
Whereas my 3090 (24GB memory) in v4.10.3 could handle a batchsize of ~32, in 4.11.3 this is reduced to ~10.

The problem arises when using:
* my own modified scripts

The tasks I am working on is:
*  ASR

## To reproduce

1. Run script with v4.10 and v4.11 and watch CUDA memory usage

Reproduce script (relatively minimal):
```
from transformers import Wav2Vec2ForCTC, Wav2Vec2Processor, TrainingArguments
from transformers.trainer import Trainer
from torch.utils.data.dataset import Dataset
import numpy as np

class ProcessedDataset(Dataset):
    def __init__(self, processor):
        self.processor = processor

    def __getitem__(self, i):
        x = np.ones(16000 * 10) # 10 seconds
        y = "this is a random sentence"
        with self.processor.as_target_processor():
            batch= {"labels": self.processor(y).input_ids}
        batch["input_values"] = self.processor(x, sampling_rate=16000).input_values
        return batch

    def __len__(self):
        return 10000

class DataCollator:
    def __init__(self, processor):
        self.processor = processor

    def __call__(self, features):
        input_features = [{"input_values": feature["input_values"][0]} for feature in features]
        label_features = [{"input_ids": feature["labels"]} for feature in features]
        batch = self.processor.pad(
            input_features,
            padding=True,
            max_length=None,
            pad_to_multiple_of=None,
            return_tensors="pt",
        )
        with self.processor.as_target_processor():
            labels_batch = self.processor.pad(
                label_features,
                padding=True,
                max_length=None,
                pad_to_multiple_of=None,
                return_tensors="pt",
            )
        labels = labels_batch["input_ids"].masked_fill(labels_batch.attention_mask.ne(1), -100)
        batch["labels"] = labels
        return batch


proc = Wav2Vec2Processor.from_pretrained("wietsedv/wav2vec2-large-xlsr-53-dutch")
model = Wav2Vec2ForCTC.from_pretrained(
    "facebook/wav2vec2-large-nl-voxpopuli",
    attention_dropout=0,
    hidden_dropout=0,
    feat_proj_dropout=0,
    mask_time_prob=0,
    layerdrop=0,
    activation_dropout=0,
    gradient_checkpointing=True,
    ctc_loss_reduction="mean",
    pad_token_id=proc.tokenizer.pad_token_id,
    vocab_size=len(proc.tokenizer),
    ctc_zero_infinity=True
)
ds = ProcessedDataset(proc)
data_collator = DataCollator(processor=proc)
args = TrainingArguments(
    output_dir="/tmp/tmp_model",
    per_device_train_batch_size=8,
    gradient_accumulation_steps=1,
    do_eval=False,
    num_train_epochs=1,
    fp16=True,
    group_by_length=False,
    save_steps=-1,
    eval_steps=1024,
    logging_steps=1024,
    warmup_steps=128,
    save_total_limit=1,
    dataloader_num_workers=1,
    seed=11
)

trainer = Trainer(model=model, args=args, train_dataset=ds, data_collator=data_collator)
trainer.train()

```

## Expected behavior

Upgrading Huggingface Transformers from 4.10 to a later version should keep the memory usage in the same ballpark

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Fixes #14388

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/modeling_utils.py [hunk 1]
```diff
@@ -1401,8 +1402,16 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
                     model = cls(config, *model_args, **model_kwargs)
         else:
             with no_init_weights(_enable=_fast_init):
+
+                restore_gradient_checkpointing = False
```
Comment (patrickvonplaten) [inline src/transformers/modeling_utils.py:1406]: should probably add some comments and a test here. @sgugger - let me know what if you think the PR would be good for you

File: src/transformers/modeling_utils.py [hunk 2]
```diff
@@ -1401,8 +1402,16 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
...
+                if hasattr(config, "gradient_checkpointing") and config.gradient_checkpointing:
+                    restore_gradient_checkpointing = True
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1408]: This is just :-)
```suggestion
                restore_gradient_checkpointing = getattr(config, "gradient_checkpointing", False)
```

File: src/transformers/modeling_utils.py [hunk 3]
```diff
@@ -1401,8 +1402,16 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
...
+
                 model = cls(config, *model_args, **model_kwargs)
 
+                if restore_gradient_checkpointing:
+                    config.gradient_checkpointing = True
+
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1414]: I don't think this is necessary, we can just keep the `restore_gradient_checkpointing` bool

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -1454,6 +1463,11 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
         # Set model in evaluation mode to deactivate DropOut modules by default
         model.eval()
 
+        if getattr(model.config, "gradient_checkpointing", False):
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1466]:
```suggestion
        if restore_gradient_checkpointing:
```

File: src/transformers/modeling_utils.py [hunk 5]
```diff
@@ -1454,6 +1463,11 @@ def from_pretrained(cls, pretrained_model_name_or_path: Optional[Union[str, os.P
...
+            model.gradient_checkpointing_enable()
+            # Remove the attribute now that is has been consumed, so it's no saved in the config.
+            delattr(model.config, "gradient_checkpointing")
```
Comment (sgugger) [inline src/transformers/modeling_utils.py:1469]:
```suggestion
```

**Review by sgugger — COMMENTED** [review]
Thanks for fixing! A small reproducer I have that can become a test is using a `from_pretrained` for a model (we could use one of the tiny models for tests):
```
model = BertModel.from_pretrained("tiny_checkpoint", gradient_checkpointing=True)
assert model.encoder.gradient_checkpointing
```

The problem comes from deeper actually, as doing the following fails:
```
config = BertConfig.from_pretrained("bert-base-cased")
config.gradient_checkpointing = True

model = BertModel(config)
assert model.encoder.gradient_checkpointing
```

Trying to see if I can fix this at the root.
